Frontend/bugfix/datepicker tz

lucaslcode opened the pull request:
closes #1068

This seems to be working but it's impossible to do tests because you can't mock the local timezone.

The problem was `Date.toISOString()` always gives the iso string in UTC time.
`dayjs().format()` gives the local iso string so I changed everything to that.

**Frontend checklist**
- [x] Formatted my code with `yarn format && yarn lint --fix`
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [ ] Added any new components to storybook
- [x] Added tests where relevant
- [ ] All tests pass
- [x] Clicked around my changes running locally and it works
- [x] Checked Desktop, Mobile and Tablet screen sizes


<!---
Remember to request review from couchers-org/frontend, couchers-org/@backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

lucaslcode committed:
add failing tests

lucaslcode committed:
change to dayjs

lucaslcode committed:
skip failing test

darrenvong commented on app/frontend/src/features/auth/signup/CompleteSignupForm.tsx:
```diff
@@ -204,8 +205,14 @@ export default function CompleteSignupForm() {
-              error={!!errors?.birthdate?.message}
-              helperText={errors?.birthdate?.message ?? " "}
+              error={
+                //@ts-ignore Dayjs type breaks this
```
What errors do you get with this? Would have thought it'd be ok with this since you are converting the value to boolean with `!!`

lucaslcode replied:
From memory, something like `DeepMap<Dayjs, FieldError> does not contain member 'message'`

darrenvong commented on app/frontend/src/components/Datepicker.test.tsx:
```diff
@@ -37,17 +29,36 @@ const Form = ({ submitForm }: { submitForm: (data: unknown) => void }) => {
+
+  //Doesn't work because it seems impossible to mock the local timezone
+  //so date.format() gives the time in utc
+  it.skip.each`
```
Will using this work? https://www.npmjs.com/package/timezone-mock

lucaslcode replied:
Didn't work, I think because mockdate already replaces the date object. I can try it without setting a mock date, though, but it will be flakey. Or maybe I can fork it and add support for a mock date as well as a mock tz.

darrenvong replied:
Probably worth pulling this in and for tests that don't need time zone mocking, mock the relevant Date methods with jest directly? They aren't as hard to mock compared to time zones

lucaslcode committed:
attempt tests

[skip ci]

lucaslcode committed:
attempt tests 2

lucaslcode commented:
@darrenvong still can't get it working. You're welcome to have a go, otherwise this might have to be untested...

darrenvong committed:
Fix Datepicker test to work with timezones

`new Date('yyyy-mm-dd')` without a time always defaults to UTC, hence
the need to add '00:00' after it

See "Timestamp string" note in
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date

darrenvong commented:
I managed to fix the tests! You were pretty close, the thing it was failing on was on the `mockDate`, since `new Date('2021-03-20')` (without time) always defaults to UTC no matter what timezone you are in, so the fix was literally adding the time at the end like "00:00" to force it to use the local time zone

darrenvong reviewed (approved):
This works when I did another sanity test for real on the browser, so let's get this in once CI passes 🎉

lucaslcode commented:
Thanks!!
